src/components/smart/Table.vue

```diff
@@ -1,69 +1,321 @@
+    (!isPrevious && page.value < props.pagination!.totalPages)
+  ) {
+    page.value += isPrevious ? -1 : 1
+  }
+
+  emit("pageNumber", page.value)
+}
+
+// The working version of the list that is used to perform operations upon
+const workingList = useVModel(props, "list", emit)
+
+// Spinner functionality
+const isSpinnerEnabled = ref(false)
+const showSpinner = (duration: number = 500) => {
+  isSpinnerEnabled.value = true
+  setTimeout(() => {
+    isSpinnerEnabled.value = false
+  }, duration)
+}
+
+watch(
+  () => props.spinner,
+  () => {
+    if (props.spinner?.enabled === true) {
+      showSpinner(props.spinner.duration)
+    }
+  },
+)
+
+// Checkbox functionality
+const selectedRows = useVModel(props, "selectedRows", emit)
+
+watch(workingList.value, (updatedList) => {
+  if (props.checkbox) {
+    updatedList = updatedList.map((item) => ({
+      ...item,
+      selected: false,
+    }))
+  }
+})
+
 const onRowClicked = (item: Item) => emit("onRowClicked", item)
+
+const isRowSelected = (item: Item) => {
+  const { selected, ...data } = item
+  return selectedRows.value?.some((row) => isEqual(row, data))
+}
+
+const toggleRow = (item: Item) => {
+  item.selected = !item.selected
+  const { selected, ...data } = item
+
+  const index = selectedRows.value?.findIndex((row) => isEqual(row, data)) ?? -1
+
+  if (item.selected && !isRowSelected(data)) selectedRows.value!.push(data)
+  else if (index !== -1) selectedRows.value?.splice(index, 1)
+}
+
+const selectAllCheckbox = ref<HTMLInputElement | null>(null)
+
+const toggleAllRows = () => {
+  const isChecked = selectAllCheckbox.value?.checked
+  workingList.value.forEach((item) => (item.selected = isChecked))
+
+  if (isChecked) {
+    workingList.value.forEach((item) => {
+      const { selected, ...data } = item
+      if (!isRowSelected(item)) selectedRows.value!.push(data)
+    })
+  } else {
+    workingList.value.forEach((item) => {
+      const { selected, ...data } = item
+      const index =
+        selectedRows.value?.findIndex((row) => isEqual(row, data)) ?? -1
+      selectedRows.value!.splice(index, 1)
+    })
+  }
+}
+
+const areAllRowsSelected = computed(() => {
+  if (workingList.value.length === 0 || selectedRows.value?.length === 0)
+    return false
+
+  let count = 0
+  workingList.value.forEach((item) => {
+    const { selected, ...data } = item
+    if (selectedRows.value?.findIndex((row) => isEqual(row, data)) !== -1) {
+      count += 1
+    }
+  })
+  return count === workingList.value.length
+})
+
+// Sort List by key and direction which can set to ascending or descending
+export type Direction = "ascending" | "descending"
+
+const sortList = (key: string, direction: Direction) => {
+  workingList.value = workingList.value?.sort((a, b) => {
+    const valueA = a[key] as string
+    const valueB = b[key] as string
+    return direction === "ascending"
+      ? valueA.localeCompare(valueB)
+      : valueB.localeCompare(valueA)
+  })
+}
```

Comment:
An explicit assignment isn't necessary here since `sort` already mutates the array. We could also follow the immutability-based approach.

```js
const sortList = (key: string, direction: Direction) => {
  const sortedList = [...workingList.value];
  sortedList.sort((a, b) => {
    const valueA = a[key] as string;
    const valueB = b[key] as string;
    return direction === "ascending"
      ? valueA.localeCompare(valueB)
      : valueB.localeCompare(valueA);
  });

  workingList.value = sortedList;
};

```